cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/adapters/ApplicationMetaAdapter.java

```diff
@@ -0,0 +1,149 @@
+  static final String APPSPEC_REDUCTION_ENABLED = "appspec.reduction";
+  private final AppSpecDeserializationContext context;
+
+  /**
+   * Constructs an adapter with a shared deserialization context.
+   */
+  public ApplicationMetaAdapter(AppSpecDeserializationContext context) {
+    this.context = context;
+  }
+
+  /**
+   * Serializes an {@link ApplicationMeta} object to its JSON representation. Includes the
+   * application ID and its specification.
+   */
+  @Override
+  public JsonElement serialize(ApplicationMeta src, Type typeOfSrc,
+      JsonSerializationContext context) {
+    JsonObject jsonObject = new JsonObject();
+    jsonObject.addProperty("id", src.getId());
+
+    JsonElement specJson = context.serialize(src.getSpec());
+    jsonObject.add("spec", specJson);
+
+    return jsonObject;
+  }
+
+  /**
+   * Deserializes a JSON element into an {@link ApplicationMeta} object. It extracts the
+   * application's {@link ArtifactId} from its specification JSON to set the parent name and
+   * namespace in the shared {@link AppSpecDeserializationContext}. This enables correct resolution
+   * of plugins within the application's scope.
+   *
+   * @throws JsonParseException If JSON is malformed or required fields are missing.
+   */
+  @Override
+  public ApplicationMeta deserialize(JsonElement json, Type typeOfT,
+      JsonDeserializationContext context) throws JsonParseException {
+    JsonObject jsonObject = json.getAsJsonObject();
+
+    JsonElement idElement = jsonObject.get("id");
+    if (idElement == null || idElement.isJsonNull()) {
+      throw new JsonParseException("ApplicationMeta 'id' field is missing or null");
+    }
+    String id = idElement.getAsString();
+
+    JsonObject specJson = jsonObject.getAsJsonObject("spec");
+    if (specJson == null || specJson.isJsonNull()) {
+      throw new JsonParseException("ApplicationMeta 'spec' field is missing or null for id: " + id);
+    }
+
+    // Set parent context before fully deserializing the AppSpec,
+    // so plugins can be retrieved from DB using this parent information.
+    JsonElement artifactIdJson = specJson.get("artifactId");
+    if (artifactIdJson != null && !artifactIdJson.isJsonNull()) {
+      ArtifactId artifactId = context.deserialize(artifactIdJson, ArtifactId.class);
+      if (artifactId != null) {
+        String parentName = artifactId.getName();
+        String parentNamespace =
+            artifactId.getScope().equals(ArtifactScope.SYSTEM) ? NamespaceId.SYSTEM.getNamespace()
+                : this.context.getNamespace();
+        this.context.setParentName(parentName);
+        this.context.setParentNamespace(parentNamespace);
+      } else {
+        throw new JsonParseException(
+            "ArtifactId in ApplicationSpecification was null for app: " + id);
+      }
+    }
+    ApplicationSpecification spec = context.deserialize(specJson, ApplicationSpecification.class);
+    return new ApplicationMeta(id, spec, null, null);
```

Comment:
How are we handling the SCM / edit details that is missed here ? or is it being added at a further stage ?